tensorboard/webapp/metrics/actions/index.ts

```diff
@@ -172,8 +173,12 @@ export const metricsShowAllPlugins = createAction(
 export const linkedTimeSelectionChanged = createAction(
   '[Metrics] Linked Time Selection Changed',
   props<{
-    startStep: number;
-    endStep: number | undefined;
+    timeSelection: {
+      startStep: number;
+      endStep: number | undefined;
+    };
+    affordance?: TimeSelectionAffordance | undefined;
+    isDragging?: boolean | undefined;
```

Comment:
Some questions/comments:

1. Why doesn't this Action use the TimeSelection type?

2. What is the difference between affordance = undefined and affordance = TimeSelectionAffordance.NONE and affordance just isn't specified (there are three ways to basically ignore affordance)? Would be nice to tighten this up so only one or two of those is possible.

3. Although it's true that affordance doesn't really matter when isDragging = true, I still think it is worth setting it because it better represents the truth of what is happening and allows you to specify a more accurate data model.

4. Can we use TypeScript to better describe and enforce the relationship between affordance and isDragging?

```
export type TimeSelectionChange = {
  timeSelection: TimeSelection;
  affordance: TimeSelectionAffordance.EXTENDED_LINE | TimeSelectionAffordance.FOB;
  isDragging: boolean;
} | {
  timeSelection: TimeSelection;
  affordance: TimeSelectionAffordance.FOB_REMOVED | TimeSelectionAffordance.FOB_TEXT | TimeSelectionAffordance.NONE;
  isDragging: undefined;
} |
// TODO: Delete this case when affordance is specified everywhere.
{
  timeSelection: TimeSelection;
  affordance: undefined;
  isDragging: undefined;
};

export const linkedTimeSelectionChanged = createAction(
  '[Metrics] Linked Time Selection Changed',
  props<TimeSelectionChange>()
);
```

This will possibly make the rest of your code a bit more complicated but could be worth it?

Reply:
Just to make sure this wasn't an unreasonable suggestion, I tried out the changes here:

https://github.com/bmd3k/tensorboard/commit/49686f2f7d0af027a6bc987bdf8428f2ee70e94c

The code actually looks a little clearer, mostly because of using TimeSelection in the action. I didn't modify the tests, though.

Reply:
For 1, it was inherited definition. The type [TimeSelectionView](https://cs.opensource.google/tensorflow/tensorboard/+/master:tensorboard/webapp/metrics/views/card_renderer/utils.ts;l=88?q=TimeSelectionView&sq=&ss=tensorflow%2Ftensorboard) is used at card level we use instead of TimeSelection. (And that's the same reason sometimes we need method like "convertToTimeSelection".) Same, we use TimeSelectionView on actions. 
I think this could be changed but not sure if that actually cause more "convert method" in between.

2-4, after thinking for a while, the type does provide clear structure, isDragging only matters when the affordance is FOB or EXTEND_LINE. Will work on the adjustment.

Reply:
We spoke offline: Defining and using isDragging properly seems to be more work than its worth at this point. We agreed to just use the affordance property. If affordance is undefined then we don't log.